### DISCUSSION TITLE
«Other» variant for discriminated unions

### DISCUSSION THREAD

**Discussion by eigenein** [pydantic/pydantic#4928] [Question]
Hi 👋,

I'm looking for a way to define a tagged union which would allow capturing undefined variants via a fallback model. The closest similar thing I can think of is the [Serde](https://serde.rs/)'s [`#[serde(other)]`](https://serde.rs/variant-attrs.html#other) in the Rust world.

Here's a pseudocode of what I'm trying to achieve:

```python
class Foo(BaseModel):
    code: Literal["foo"]


class Other(BaseModel):
    code: str


class Parent(BaseModel):
    foobar: Annotated[Union[Foo, Other], Field(discriminator="code")]


# This would validate the payload against `Foo` since `code="foo"`:
assert Parent.parse_obj({"foobar": {"code": "foo"}}).foobar == Foo(code="foo")

# This would validate against `Other` since `code` is something undefined:
assert Parent.parse_obj({"foobar": {"code": "quuuux"}}).foobar == Other(code="quuuux")
```

This surely doesn't work because `pydantic` reasonably expects `Other.foo` to be annotated with a `Literal`. But is there a way to make it work?

Reasoning: an API returns an error code (which would be my discriminator), and I want to define a separate model per error code (because payloads for different errors differ as well), but I want to capture unexpected error codes under some `class OtherError(BaseModel)` (the last thing I want is a `ValidationError` while parsing a legit API error).

**Comment by samuelcolvin** [comment]
No easy way in v1.x, but in V2 the discriminator can be a function so this should be easy.

**Reply by austinorr** [reply]
@samuelcolvin I'm having trouble finding this in the V2 docs, did 'discriminator function' actually make it to V2? Maybe it's called something new and I just don't know what to search.

**Reply by eigenein** [reply]
Now, it is here: https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions-with-callable-discriminator

**Reply by eigenein** [reply]
Although, I don't quite get how the callable `Discriminator` would be helpful here. Could you elaborate?

**Reply by Defman21** [reply]
> Although, I don't quite get how the callable `Discriminator` would be helpful here. Could you elaborate?

The only way I've found to use `pydantic.Discriminator` to achieve this (without the Union first-one-matched hack) is something like this:

```python
import typing

from pydantic import BaseModel, Tag, Discriminator


class Cat(BaseModel):
    name: str
    type: typing.Literal['cat']


class Dog(BaseModel):
    name: str
    type: typing.Literal['dog']


class Animal(BaseModel):
    name: str
    type: str


def animal_discriminator(v: typing.Any):
    if isinstance(v, dict):
        type_ = v.get('type', None)
    else:
        type_ = getattr(v, 'type', None)

    if type_ in {'cat', 'dog'}:
        return type_
    return 'animal'


class Human(BaseModel):
    pet: typing.Annotated[
        typing.Union[
            typing.Annotated[Cat, Tag('cat')],
            typing.Annotated[Dog, Tag('dog')],
            typing.Annotated[Animal, Tag('animal')],
        ],
        Discriminator(animal_discriminator)
    ]


print(Human.model_validate({'pet': {'type': 'cat', 'name': 'Kitty'}}))
# pet=Cat(name='Kitty', type='cat')
print(Human.model_validate({'pet': {'type': 'dog', 'name': 'Doggy'}}))
# pet=Dog(name='Doggy', type='dog')
print(Human.model_validate({'pet': {'type': 'parrot', 'name': 'Parrot'}}))
# pet=Animal(name='Parrot', type='parrot')
```

The solution requires to keep a set of Literal values for every model you want to discriminate against, though.

**Comment by eigenein** [comment]
Looks like I did find a workaround, abusing the `Union`'s «first one matched» functionality:

```python
class Cat(BaseModel):
    t: Literal["cat"]
    

class Dog(BaseModel):
    t: Literal["dog"]
    

class Other(BaseModel):
    t: str
    

class Container(BaseModel):
    foo: Annotated[Union[Cat, Dog], Field(discriminator="t")] | Other
```

```python
>>> Container(foo={"t": "cat"})
Container(foo=Cat(t='cat'))

>>> Container(foo={"t": "dog"})
Container(foo=Dog(t='dog'))

>>> Container(foo={"t": "wtf"})
Container(foo=Other(t='wtf'))
```

**Reply by spazm** [reply]
The default union type has changed from "left-to-right" to "smart_mode", so for your outer union we can force left-to-right with `= Field(union_mode='left_to_right')`

```python
class Container(BaseModel):
    foo: Annotated[Union[Cat, Dog], Field(discriminator="t")] | Other = Field(union_mode="left_to_right")
```

**Reply by spazm** [reply]
```python
CatDogOther = Annotated[
    Union[
        Annotated[Union[Cat, Dog],  Field(discriminator="t")],
        Other
  ], Field(union_mode="left_to_right")
]

class Container(BaseModel):
    foo: CatDogOther
```
